web/ww.ts

```diff
@@ -135,215 +270,106 @@ class Wormhole {
-
-					const offer = await this.pc.createOffer();
-					console.log("created offer");
-					this.ws.send(webwormhole.seal(this.key, JSON.stringify(offer)));
-					this.state = "wait_for_webtc_answer";
-					this.pc.setLocalDescription(offer);
-				});
+	makePeerConnection(iceServers: GoICEServers): RTCPeerConnection {
+		let normalisedICEServers = [];
+		for (let i = 0; i < iceServers.length; i++) {
+			normalisedICEServers.push({
+				urls: iceServers[i].URLs,
+				username: iceServers[i].Username,
+				credential: iceServers[i].Credential,
+			});
+		}
+		const pc = new RTCPeerConnection({
+			iceServers: normalisedICEServers,
+		});
+		pc.onicecandidate = (e: RTCPeerConnectionIceEvent) => {
+			if (!this.ws || !this.key || !this.pc) {
 				return;
 			}
 
-			case "wait_for_pake_b": {
-				console.log("got pake message b:", m.data);
-				this.key = webwormhole.finish(m.data);
-				if (!this.key) {
-					this.fail("could not generate key");
-					return;
-				}
-				console.log("generated key");
-				this.state = "wait_for_webtc_offer";
-				return;
+			if (e.candidate && e.candidate.candidate !== "") {
+				console.log("got local candidate", e.candidate.candidate);
+				this.ws.send(webwormhole.seal(this.key, JSON.stringify(e.candidate)));
 			}
+		};
+		return pc;
+	}
 
-			case "wait_for_webtc_offer": {
-				if (!this.key || !this.pc) { return }
+	async connType(): Promise<string> {
+		if (!this.pc) {
+			return "";
+		}
 
-				const msg: RTCSessionDescriptionInit | null = JSON.parse(webwormhole.open(this.key, m.data));
-				if (!msg) {
-					this.fail("bad key");
-					this.ws.send(webwormhole.seal(this.key, "bye"));
-					this.ws.close(WormholeErrorCodes.closeBadKey);
-					return;
-				}
-				if (msg.type !== "offer") {
-					console.log("unexpected message", msg);
-					this.fail("unexpected message");
-					return;
-				}
-				console.log("got offer");
-				// No intermediate state wait_for_pc_initialize because candidates can
-				// staring arriving straight after the offer is sent.
-				this.state = "wait_for_candidates";
-				this.promise2.then(async () => {
-					if (!this.key || !this.pc) { return }
-
-					await this.pc.setRemoteDescription(new RTCSessionDescription(msg));
-					const answer = await this.pc.createAnswer();
-					console.log("created answer");
-					this.ws.send(webwormhole.seal(this.key, JSON.stringify(answer)));
-					this.resolve3(webwormhole.fingerprint(this.key));
-					this.pc.setLocalDescription(answer);
-				});
-				return;
+		// RTCStatsReport.forEach is all that's defined in the TypeScript DOM defs, which
+		// makes this kind of awkward. Ah well.
+		const stats = await this.pc.getStats();
+
+		let id: string | undefined;
+		stats.forEach((s: RTCIceCandidatePairStats) => {
+			// s.selected gives more confidenece than s.state == "succeeded", but Chrome does
+			// not implement it.
+			if (
+				s.type === "candidate-pair" &&
+				(s as RTCIceCandidatePairStats).state === "succeeded"
```

Comment:
the point of line 327 suggestion was that now this casting is not needed @saljam

Reply:
indeed. however, having had another look even that doesn't look right to me: not every RTCStats is an RTCIceCandidatePairStats.

in fact, the second forEach call (syntax error aside) should take a RTCIceCandidateStats and not a RTCIceCandidatePairStats. but RTCIceCandidateStats is not in the typescript dom library.

sounds like the best approach is to accept RTCStats in the callback, and do the appropriate type assertion for each value based on s.type. i'll revert this for now.